google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/models.scala

```diff
@@ -6,9 +6,33 @@ package akka.stream.alpakka.googlecloud.pubsub
 
 import java.time.Instant
 
+import akka.actor.ActorSystem
+import akka.annotation.InternalApi
+import akka.http.scaladsl.{Http, HttpExt}
+import akka.stream.alpakka.googlecloud.pubsub.impl.{GoogleSession, GoogleTokenApi}
+
 import scala.collection.immutable
 import scala.collection.JavaConverters._
 
+private[pubsub] class PubSubConfig(val projectId: String,
```

Comment:
Wow! I saw this as a pattern somewhere, and I could use this in the docs.examples package without a problem. If the val-s are protected by default I think thats not a problem.
I think the common user will never want to use this class. Just create one instance with the given apply/create and pass it as a value.

Reply:
The package private `private[...]` does not apply when used from Java. So maybe that usage you noticed was from Java.

Reply:
Okay! Today I played with it, and I think it's not worth it. If I make that constructor private I can't tast the code. I can't extend from that class anymore so I can't change the session to a mock session.

On the other hand I tried to use the new keyword from java. For that, you need a `Http.apply(actorSystem)` which is not really compile, so if you really want to hack through yourself with the constructor you need to do something like `Http.createExtension((ExtendedActorSystem) system)` which is really ugly and I think it's rings some bells in any developers head.

Do you have any other idea?

Reply:
Sooo we open up a protected/package-private builder method just for the testing, and with that we could hide the constructor.

[Here](https://github.com/akka/alpakka/pull/1258/files/1eb96f5f1c41ffd9a6ed4679cafaf993a9499f9f#diff-adb757a87b9b625e66a81647d57b68abR40) if we use an apply and a protected(?) withSession the concept will work?

Reply:
Hm. It does not look the best, I agree. Lets be explicit why providing custom session is public by creating a "testkit" factory, that would allow custom session parameter. Similar to the pattern in the [reference connector](https://github.com/akka/alpakka/blob/master/reference/src/main/scala/akka/stream/alpakka/reference/testkit/MessageFactory.scala).

Reply:
Okay;

```
class PubSubConfig private (val projectId: String,
                            val apiKey: String,
                            clientEmail: String,
                            privateKey: String,
                            inSession: GoogleSession) {
  @InternalApi
  protected[pubsub] lazy val session = inSession
}

object PubSubConfig {
  protected[pubsub] def apply(projectId: String,
                              apiKey: String,
                              clientEmail: String,
                              privateKey: String,
                              session: GoogleSession) =
    new PubSubConfig(projectId, apiKey, clientEmail, privateKey, session)
...
```

Works halfway... Now the constructor is private, but if I write down the val keyword before session, an auto getter is generated which will leak the session (java only). If this is not a problem, I can push my code (without the session reassign).

Reply:
Ahh. Changing to `private val session` should do it.

Reply:
But if its private I can't access it from the pubsub package. So the circle is closed :( I need it to be unseeable, unmodifyable from outside the pubsub package, but I want to use it as a normal parameter inside. I think java can't do this, or I still didn't see the full picture... Either way I would be pleased if you would try to play with this for a bit and probably commit a "suitable way of doing this" example, so in the future I can use that as a reference.